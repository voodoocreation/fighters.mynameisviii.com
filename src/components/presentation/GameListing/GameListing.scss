@import "../../../scss/variables", "../../../scss/mixins";

.GameListing {
  flex-basis: 100%;
  padding: $spacing-quarter;

  @include screen-above($screen-sm-min) {
    flex-basis: 50%;
  }
}
