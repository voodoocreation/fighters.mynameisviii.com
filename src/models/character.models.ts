import { DeepPartial } from "redux";

import { createSlugFromString } from "../helpers/dataTransformers";
import { IMoveset, moveset } from "./moveset.models";

export interface ICharacter {
  imageUrl: string;
  movesets: IMoveset[];
  name: string;
  slug: string;
}

export const character = (
  options: DeepPartial<ICharacter> = {}
): ICharacter => ({
  imageUrl: options.imageUrl || "",
  movesets: options.movesets ? options.movesets.map(moveset) : [],
  name: options.name || "",
  slug: options.slug || createSlugFromString(options.name || "")
});
