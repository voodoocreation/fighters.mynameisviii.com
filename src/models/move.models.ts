import { DeepPartial } from "redux";

export interface IMove {
  combo: string;
  conditions?: string;
  name: string;
}

export const move = (options: DeepPartial<IMove> = {}): IMove => ({
  combo: options.combo || "",
  conditions: options.conditions || undefined,
  name: options.name || ""
});
