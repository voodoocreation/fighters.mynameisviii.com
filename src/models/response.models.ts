export interface ISuccess<Data> {
  ok: true;
  data: Data;
}

export const success = <T>(data: T): ISuccess<T> => ({ ok: true, data });

export interface IFailure {
  ok: false;
  message: string;
}

export const failure = (message: string): IFailure => ({ ok: false, message });
