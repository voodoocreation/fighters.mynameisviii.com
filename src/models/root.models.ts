export * from "./character.models";
export * from "./error.models";
export * from "./game.models";
export * from "./move.models";
export * from "./moveset.models";
export * from "./response.models";
