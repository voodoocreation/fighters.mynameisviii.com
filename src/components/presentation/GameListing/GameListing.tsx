import * as React from "react";
import { InjectedIntlProps, injectIntl } from "react-intl";

import { IGame } from "../../../models/game.models";
import Link from "../Link/Link";

import "./GameListing.scss";

interface IProps extends IGame, InjectedIntlProps {
  isSelected?: boolean;
  onClick: () => void;
}

class GameListing extends React.Component<IProps> {
  public static defaultProps = {
    isSelected: false
  };

  public render() {
    const { imageUrl, isSelected, title } = this.props;
    const { formatMessage } = this.props.intl;

    return (
      <article className="GameListing">
        <Link
          route={!isSelected ? `/games/${this.props.slug}` : undefined}
          onClick={this.props.onClick}
        >
          <img
            src={imageUrl}
            alt={formatMessage(
              { id: "VIEW_ALL_MOVES_FOR_GAME" },
              { game: title }
            )}
          />
        </Link>
      </article>
    );
  }
}

const GameListingWrapped = injectIntl(GameListing);

export default GameListingWrapped;
