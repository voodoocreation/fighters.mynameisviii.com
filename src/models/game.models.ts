import { DeepPartial } from "redux";

import { createSlugFromString } from "../helpers/dataTransformers";
import { character, ICharacter } from "./character.models";

export interface IGame {
  imageUrl: string;
  slug: string;
  title: string;
  characters: ICharacter[];
}

export const game = (options: DeepPartial<IGame> = {}): IGame => ({
  characters: options.characters ? options.characters.map(character) : [],
  imageUrl: options.imageUrl || "",
  slug: options.slug || createSlugFromString(options.title || ""),
  title: options.title || ""
});
