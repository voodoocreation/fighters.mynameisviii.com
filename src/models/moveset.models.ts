import { DeepPartial } from "redux";

import { IMove, move } from "./move.models";

export interface IMoveset {
  moves: IMove[];
  name: string;
}

export const moveset = (options: DeepPartial<IMoveset> = {}): IMoveset => ({
  moves: options.moves ? options.moves.map(move) : [],
  name: options.name || "",
});
