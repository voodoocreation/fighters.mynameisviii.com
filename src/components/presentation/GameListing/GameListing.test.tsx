import { character, game } from "../../../models/root.models";
import ComponentTester from "../../../utilities/ComponentTester";

import GameListing from "./GameListing";

const testGame = game({
  characters: [
    character({
      name: "Test character 1"
    }),
    character({
      name: "Test character 2"
    })
  ],
  imageUrl: "imageUrl",
  slug: "test-game",
  title: "Test Game"
});

const component = new ComponentTester(GameListing).withDefaultProps({
  ...testGame,
  onClick: jest.fn()
});

describe("[presentation] <GameListing />", () => {
  describe("when the listing isn't selected", () => {
    const { actual, props } = component.mount();

    beforeAll(() => {
      jest.clearAllMocks();
    });

    it("renders without the isSelected class", () => {
      expect(actual.find(".isSelected")).toHaveLength(0);
    });

    it("renders correct link route", () => {
      expect(actual.find("Link").prop("route")).toBe("/games/test-game");
    });

    describe("when clicking the listing", () => {
      it("clicks the link", () => {
        actual.find("Link").simulate("click");
      });

      it("calls the onClick prop", () => {
        expect(props.onClick).toHaveBeenCalledTimes(1);
      });
    });

    it("matches snapshot", () => {
      expect(actual.render()).toMatchSnapshot();
    });
  });

  describe("when the listing is selected", () => {
    const { actual, props } = component
      .withProps({
        isSelected: true
      })
      .mount();

    beforeAll(() => {
      jest.clearAllMocks();
    });

    it("renders with the isSelected class", () => {
      expect(actual.find(".isSelected")).toHaveLength(1);
    });

    it("renders correct link route", () => {
      expect(actual.find("Link").prop("route")).toBeUndefined();
    });

    describe("when clicking the listing", () => {
      it("clicks the link", () => {
        actual.find("Link").simulate("click");
      });

      it("calls the onClick prop", () => {
        expect(props.onClick).toHaveBeenCalledTimes(1);
      });
    });

    it("matches snapshot", () => {
      expect(actual.render()).toMatchSnapshot();
    });
  });
});
