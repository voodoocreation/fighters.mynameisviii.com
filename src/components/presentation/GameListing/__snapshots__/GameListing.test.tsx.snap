// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <GameListing /> when the listing is selected matches snapshot 1`] = `
<article
  class="GameListing"
>
  <span>
    <img
      alt="View all moves for Test Game"
      src="imageUrl"
    />
  </span>
</article>
`;

exports[`[presentation] <GameListing /> when the listing isn't selected matches snapshot 1`] = `
<article
  class="GameListing"
>
  <a
    href="/games/test-game"
  >
    <img
      alt="View all moves for Test Game"
      src="imageUrl"
    />
  </a>
</article>
`;
